src/transformers/trainer_tf.py

```diff
@@ -599,7 +656,8 @@ def training_step(self, features, labels):
         Subclass and override to inject some custom behavior.
         """
         per_example_loss, _ = self.run_model(features, labels, True)
-        scaled_loss = per_example_loss / self.total_train_batch_size
+        nb_instances_in_global_batch = tf.cast(features["nb_instances_in_batch"][0], dtype=per_example_loss.dtype)
```

Comment:
A nicer solution would be to replace `features["nb_instances_in_batch"][0]` by `self.nb_instances_in_batch` and compute this value during the creation of the dataset in `get_train_tfdataset`, `get_eval_tfdataset` and `get_test_tfdataset`

Reply:
> 
> 
> A nicer solution would be to replace `features["nb_instances_in_batch"][0]` by `self.nb_instances_in_batch` and compute this value during the creation of the dataset in `get_train_tfdataset`, `get_eval_tfdataset` and `get_test_tfdataset`

This value `nb_instances_in_batch` is not a constant, and since we use shuffle() and repeat() in the training dataset, I am not sure how we can precompute it in `get_train_tfdataset` and store it in `self.nb_instances_in_batch`.

        ds = (
            self.train_dataset.repeat()
            .shuffle(self.num_train_examples, seed=self.args.seed)
            .batch(self.total_train_batch_size, drop_remainder=self.args.dataloader_drop_last)
            .prefetch(tf.data.experimental.AUTOTUNE)
        )

Reply:
Indeed. What about to compute it on the fly? I mean just before computing the scaled loss, basically as it is done in the loss function.

Reply:
> 
> 
> Indeed. What about to compute it on the fly? I mean just before computing the scaled loss, basically as it is done in the loss function for removing the tokens that have a -100 label value.

Compute it on the fly - before computing the scaled loss - will only work in a single replica and in a single gradient accumulation settings. Because we want to update the model's parameters by using a `global batch`, but when we enter `training_step`, it is already in a single replica context, and the batch received is just a partial batch.

Reply:
Sorry badly expressed myself, by just before I wanted to say directly in the train function before to call `distributed_training_steps`. Like this you get this number for each global batch size (train_batch*n_replicas*accumulation) and then get the same number doesn't matter if you are in distributed training or with more than one accumulation. No?

Reply:
We can compute it before the call `distributed_training_steps`, but it has to be passed to the replicas via `self.args.strategy.run(self.apply_gradients, batch, new argument)`  - I am ok with it though.

Since the current main concern is about `tokens with -100 label value`, this is a way to fix the problem, which will only modify `tf.trainer` file.

However, since we have the separation of dataset / trainer / loss-computation-in-model-class, I don't know if it is OK to make the assumption about `-100` in `tf.trainer`. But if you think it is OK, we can do it.

Reply:
Yes, please, looks the perfect solution to me!

Reply:
The usage of something else than -100 is currently deprecated so you can be safe for that.

Reply:
There is a potential subtlety here. Since we use 

    self.args.strategy.experimental_distribute_dataset(ds),

depending on the `strategy` and the hardware (single/multi CPU/GPU/TPU) settings, the `batch` in `self.distributed_training_steps` could be usual tensors, or could be packed in `PerReplica` object, something like

	PerReplica:{
	  0: <tf.Tensor: shape=(2,), dtype=int32, numpy=array([57, 94], dtype=int32)>,
	  1: <tf.Tensor: shape=(2,), dtype=int32, numpy=array([22, 70], dtype=int32)>,
	  2: <tf.Tensor: shape=(2,), dtype=int32, numpy=array([10, 77], dtype=int32)>,
	  3: <tf.Tensor: shape=(2,), dtype=int32, numpy=array([64, 82], dtype=int32)>,
	  4: <tf.Tensor: shape=(1,), dtype=int32, numpy=array([22], dtype=int32)>,
	  5: <tf.Tensor: shape=(0,), dtype=int32, numpy=array([], dtype=int32)>,
	  6: <tf.Tensor: shape=(0,), dtype=int32, numpy=array([], dtype=int32)>,
	  7: <tf.Tensor: shape=(0,), dtype=int32, numpy=array([], dtype=int32)>
	}

This is what I had (for a personal project) when I used TPU v3-8. However, I don't know if there are other types of distributed objects when we use different strategy.

@jplu , do you think it is good to assume it is either a tuple of 2 tensors, or a tuple of 2 perplica objects? In both cases, a tuple will be a (features, labels) pair.

Reply:
Exactly! The PerReplica case occurs when you have at least 2 replicas. It is better to handle both cases (with PerReplica and with usual tensor) like it is done in the prediction loop https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py#L335

Reply:
Or even simpler you can use a class field that should do the work and then no need anymore to handle those cases.

Reply:
> 
> 
> Or even simpler you can use a class field that should do the work and then no need anymore to handle those cases.

Could you elaborate the idea more? I don't know what you mean of using a class field in this case.

Reply:
Sorry, let me explain.

In the `train` function you can have something:
```
self.epoch_logging = epoch_iter - 1 + (step + 1) / self.steps_per_epoch
self.nb_instances_in_global_batch = compute_number()
self.distributed_training_steps(batch)
```

And then in the `training_step` function:
```
per_example_loss, _ = self.run_model(features, labels, True)
scaled_loss = per_example_loss / self.nb_instances_in_global_batch 
```

Reply:
But `nb_instances_in_global_batch` is not a constant, it actually differ from one batch to another batch

Reply:
Exactly, this is why the value is recomputed at each step and then used later for the computation of each step. Otherwise don't worry go for passing with a parameter, it is the same anyway, just as you said take care if we are in singe of multi replica environment.

Reply:
No problem. But just a reamk (my thought) - using class filed, and recompute for each batch won't work. Because the computational is compiled due to

```
    @tf.function
    def distributed_training_steps(self, batch):
```
And since 

    scaled_loss = per_example_loss / self.nb_instances_in_global_batch 

is included in that graph, the graph will have the initial value of `self.nb_instances_in_global_batch` the first time it is compiled. Even we recompute it outside the graph, the value is still the initial value in the graph.

This is tricky, but I do have an support example for this claim, if you want to see.

Reply:
No worries, do with a new parameter in `distributed_training_steps` it is even nicer^^